core/utils/managed_allocator.hpp

```diff
@@ -22,79 +22,93 @@
+    std::is_nothrow_copy_constructible_v<A>)
+    : manager_{other.manager_}, allocator_{other.allocator_} {}
 
-  ManagedAllocator(const ManagedAllocator& other) noexcept
-    : Allocator(other), rm_(&other.ResourceManager()) {}
+  // TODO(MBkkt) Assign:
+  //  Is it safe in case of other == &this? Looks like yes
+  //  Maybe swap idiom?
 
-  ManagedAllocator& operator=(ManagedAllocator&& other) noexcept {
-    static_cast<Allocator&>(*this) = std::move(static_cast<Allocator&>(other));
-    rm_ = &other.ResourceManager();
+  ManagedAllocator& operator=(ManagedAllocator&& other) noexcept(
+    std::is_nothrow_move_assignable_v<A>) {
+    manager_ = other.manager_;
+    allocator_ = std::move(other.allocator_);
     return *this;
   }
 
-  ManagedAllocator& operator=(const ManagedAllocator& other) noexcept {
-    static_cast<Allocator&>(*this) = other;
-    rm_ = &other.ResourceManager();
+  ManagedAllocator& operator=(const ManagedAllocator& other) noexcept(
+    std::is_nothrow_copy_assignable_v<A>) {
+    manager_ = other.manager_;
+    allocator_ = other.allocator_;
     return *this;
   }
 
-  template<typename A>
-  ManagedAllocator(const ManagedAllocator<A, Manager>& other) noexcept
-    : Allocator(other.RawAllocator()), rm_(&other.ResourceManager()) {}
+  template<typename T>
+  ManagedAllocator(const ManagedAllocator<M, T, SafeSwap>& other) noexcept(
+    std::is_nothrow_constructible_v<A, const T&>)
+    : manager_{&other.Manager()}, allocator_{other.Allocator()} {}
 
   value_type* allocate(size_type n) {
-    rm_->Increase(sizeof(value_type) * n);
+    manager_->Increase(sizeof(value_type) * n);
     Finally cleanup = [&]() noexcept {
-      rm_->DecreaseChecked(sizeof(value_type) * n);
+      manager_->DecreaseChecked(sizeof(value_type) * n);
     };
-    auto* res = Allocator::allocate(n);
+    auto* p = allocator_.allocate(n);
     n = 0;
-    return res;
+    return p;
   }
 
   void deallocate(value_type* p, size_type n) noexcept {
-    Allocator::deallocate(p, n);
-    rm_->Decrease(sizeof(value_type) * n);
+    allocator_.deallocate(p, n);
+    IRS_ASSERT(n != 0);
+    manager_->Decrease(sizeof(value_type) * n);
   }
 
-  const Allocator& RawAllocator() const noexcept {
-    return static_cast<const Allocator&>(*this);
-  }
+  M& Manager() const noexcept { return *manager_; }
+  const A& Allocator() const noexcept { return allocator_; }
 
-  template<typename A>
-  bool operator==(const ManagedAllocator<A, Manager>& other) const noexcept {
-    return RawAllocator() == other.RawAllocator() && &rm_ == &other.rm_;
+  template<typename T>
+  bool operator==(
+    const ManagedAllocator<M, T, SafeSwap>& other) const noexcept {
+    return manager_ == &other.Manager() && allocator_ == other.Allocator();
   }
 
-  Manager& ResourceManager() const noexcept { return *rm_; }
-
  private:
-  Manager* rm_;
+  M* manager_;
+  IRS_NO_UNIQUE_ADDRESS A allocator_;
 };
-}  // namespace irs
```

Comment:
Please always put last empty line to the file.
Otherwise git makes history and diff worse
I use githook.